packages/mainmenu-extension/src/builder.ts

```diff
@@ -0,0 +1,287 @@
+function dataToMenu(
+  item: ISettingRegistry.IMenu,
+  menuFactory: (options: IMainMenu.IMenuOptions) => JupyterLabMenu
+): JupyterLabMenu {
+  const { id, label, rank } = item;
+  const menu = menuFactory({ id, label, rank });
+  item.items
+    ?.filter(item => !item.disabled)
+    .sort((a, b) => (a.rank ?? Infinity) - (b.rank ?? Infinity))
+    .map(item => {
+      addItem(item, menu, menuFactory);
+    });
+  return menu;
+}
+
+function addItem(
+  item: ISettingRegistry.IMenuItem,
+  menu: JupyterLabMenu,
+  menuFactory: (options: IMainMenu.IMenuOptions) => JupyterLabMenu
+) {
+  const { command, args, submenu, type, rank } = item;
+  // Commands may not have been registered yet; so we don't force it to exist
+  menu.addItem({
+    command,
+    args: args as any,
+    submenu: submenu ? dataToMenu(submenu, menuFactory) : null,
+    type,
+    rank
+  });
+}
+
+function updateMenus(
+  menus: JupyterLabMenu[],
+  data: ISettingRegistry.IMenu[],
+  menuFactory: (options: IMainMenu.IMenuOptions) => JupyterLabMenu
+): JupyterLabMenu[] {
+  const newMenus: JupyterLabMenu[] = [];
+  data.forEach(item => {
+    const menu = menus.find(menu => menu.id === item.id);
+    if (menu) {
+      mergeMenus(item, menu, menuFactory);
+    } else {
+      newMenus.push(dataToMenu(item, menuFactory));
+    }
+  });
+  return newMenus;
+}
+
+function mergeMenus(
+  item: ISettingRegistry.IMenu,
+  menu: JupyterLabMenu,
+  menuFactory: (options: IMainMenu.IMenuOptions) => JupyterLabMenu
+) {
+  if (item.disabled) {
+    menu.dispose();
+  } else {
+    item.items?.forEach(entry => {
+      const existingItem = menu?.items.find(
+        (i, idx) =>
+          i.type === entry.type &&
+          i.command === entry.command &&
+          i.submenu?.id === entry.submenu?.id
+      );
+
+      if (existingItem && entry.type !== 'separator') {
+        if (entry.disabled) {
+          menu.removeItem(existingItem);
+        } else {
+          switch (entry.type ?? 'command') {
+            case 'command':
+              if (entry.command) {
+                if (!JSONExt.deepEqual(existingItem.args, entry.args ?? {})) {
+                  addItem(entry, menu, menuFactory);
+                }
+              }
+              break;
+            case 'submenu':
+              if (entry.submenu) {
+                mergeMenus(
+                  entry.submenu,
+                  existingItem.submenu as JupyterLabMenu,
+                  menuFactory
+                );
+              }
+          }
+        }
+      } else {
+        addItem(entry, menu, menuFactory);
+      }
+    });
+  }
+}
+
+async function displayInformation(trans: TranslationBundle): Promise<void> {
+  const result = await showDialog({
+    title: trans.__('Information'),
+    body: trans.__(
+      'Menu customization has changed. You will need to reload JupyterLab to see the changes.'
+    ),
+    buttons: [
+      Dialog.cancelButton(),
+      Dialog.okButton({ label: trans.__('RELOAD') })
```

Comment:
Maybe title case for it to be consistent?

```suggestion
      Dialog.okButton({ label: trans.__('Reload') })
```

Reply:
:+1: